File: 2015/day01-2.hs
import System.IO

solution :: String -> Int
solution content = helper 0 1 content

helper :: Int -> Int -> String -> Int
helper current index (x:xs)
    | current == 0 && x == ')' = index
    | x == '(' = helper (current + 1) (index + 1) xs
    | otherwise = helper (current - 1) (index + 1) xs

main :: IO ()
main = do
    handle <- openFile "day01.dat" ReadMode
    content <- hGetContents handle
    print . show . solution $ content

File: 2015/day01-2.ts
function solution(content: string) {
  console.log(
    content
      .split("")
      .map((current) => (current === "(" ? 1 : -1))
      .reduce((acc, cur, index) => {
        if (index === 0) return [cur];
        return [...acc, acc[index - 1] + cur];
      }, [] as number[])
      .findIndex((item) => item === -1) + 1
  );
}

Deno.readTextFile("day01.dat")
  .then((content) => content.trim())
  .then(solution);
